### PR TITLE
Improve uninitialised variable analysis for statements with on-fail

### LINKED ISSUE #38530: [Bug]: Use of `on fail` can lead to use of an uninitilized variable at runtime
### Description

When a variable is only initialized in a `transaction` or `do` block and errors are caught by an `on fail` block ballerina will assume the variable is always initialized after the `on fail` block even though the variable may not be initilized in case of an error.

At runtime this will either result in a null pointer exception (the good case actually) or the use of default values (e.g. 0 for integers) which can lead to very hard to detect bugs!

Instead, the compiler should throw a compilation error for possibly undefined variables.

### Steps to Reproduce

```ballerina
    int resultInt;
    transaction {
        resultInt = check maybeProvideInt(true);
        check commit;
    } on fail {
        io:println("Failed to initialize resultInt");
    }
    resultInt += 1;
    io:println(resultInt);  // should always print 42 but can output 1!
```

<details>

<summary>Full example code to reproduce issue (click to expand)</summary>

```ballerina
import ballerina/io;

type TestObject record {int number;};

public isolated function maybeProvideObject(boolean throwError) returns TestObject|error {
    if throwError {
        return error("Demo error.");
    }
    return {number: 42};
}

public isolated function maybeProvideInt(boolean throwError) returns int|error {
    if throwError {
        return error("Demo error.");
    }
    return 42;
}

public isolated function testUndefinedVariable() returns error? {
    int resultInt;
    transaction {
        resultInt = check maybeProvideInt(true);
        check commit;
    } on fail {
        io:println("Failed to initialize resultInt");
    }
    resultInt += 1;
    io:println(resultInt);  // should always print 42 but can output 1!

    TestObject resultObject;
    do {
        resultObject = check maybeProvideObject(true);
    } on fail {
        io:println("Failed to initialize resultObject");
    }
    io:println(resultObject.number); // throws null pointer exception
}

public function main() {
    error? err = testUndefinedVariable();
    io:println(err);
}
```

</details>

### Affected Version(s)

Tested on Ballerina 2201.2.2 (Swan Lake Update 2) and Ballerina 2201.2.0 (Swan Lake Update 2)

### OS, DB, other environment details and versions

Ubuntu in WSL 2
Language specification 2022R3
Update Tool 1.3.10

### Related area

-> Compilation

### Related issue(s) (optional)

Possibly related problem: #37854

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by pcnfernando**
Fixes #38530

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause-negative-v2.bal [hunk 1]
```diff
@@ -213,3 +213,620 @@ function testUnInitVars8(int[] data) returns string {
...
+        } else {
+            i = [check getErrorOrInt()];
+        }
+        break;
+    } on fail {
+        i = [];
+    }
+    _ = i[0]; //no compilation error
+}
+
```
Comment (pcnfernando) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause-negative-v2.bal:829]: Identified an issue in the `while` analysis logic. Working on it

Comment (pcnfernando) [reply tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause-negative-v2.bal:829]: Handled it in 2a356c593f481a5f16976a3ba60d8d0e7914757a

File: tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause-negative-v2.bal [hunk 2]
```diff
@@ -213,3 +213,837 @@ function testUnInitVars8(int[] data) returns string {
...
+        }
+    } on fail {
+        j = 1;
+    }
+    _ = j; //error: variable 'j' may not have been initialized
+}
+
+function getErrorOrIntArr() returns int[]|error {
+    return getError();
+}
```
Comment (KavinduZoysa) [inline tests/jballerina-unit-test/src/test/resources/test-src/statements/onfail/on-fail-clause-negative-v2.bal:1049]: Shall we add few destructuring tests like this.
```
import ballerina/io;
public function main() returns error? {
    int i;
    int j;
    do {
        check error("error");
        [i, j] = [1, 2];
    } on fail {
        io:println("on fail");
    }
    io:println(i); 
    io:println(j);
}
```

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java [hunk 3]
```diff
@@ -2512,10 +2641,11 @@ private void checkAssignment(BLangExpression varRef) {
 
         BSymbol symbol = ((BLangVariableReference) varRef).symbol;
         if (this.possibleFailureReached && this.uninitializedVars.containsKey(symbol)) {
-            this.uninitializedVars.put(symbol, InitStatus.PARTIAL_INIT);
-        } else {
-            this.uninitializedVars.remove(symbol);
+            getPossibleFailureUnInitVars().put(symbol, InitStatus.PARTIAL_INIT);
+        } else if (!this.possibleFailureUnInitVars.isEmpty() && !this.possibleFailureReached) {
+            getPossibleFailureUnInitVars().remove(symbol);
```
[the remark this answers is not in the document]
Comment (pcnfernando) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java:2646]: If `possibleFailureUnInitVars` is not empty, getPossibleFailureUnInitVars() wont return null

**Comment by pcnfernando** [comment]
Will be reopening a new PR once the existing blocker is resolved

**Pull request closed without merging**